packages/multichain/src/middlewares/MultichainSubscriptionManager.ts

```diff
@@ -0,0 +1,162 @@
+  method: 'eth_subscription';
+  params: {
+    subscription: Hex;
+    result: unknown;
+  };
+};
+
+type SubscriptionKey = {
+  scope: ExternalScopeString;
+  origin: string;
+  tabId?: number;
+};
+type SubscriptionEntry = SubscriptionKey & {
+  subscriptionManager: SubscriptionManager;
+};
+
+type MultichainSubscriptionManagerOptions = {
+  findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+  getNetworkClientById: NetworkController['getNetworkClientById'];
+};
+
+export class MultichainSubscriptionManager extends SafeEventEmitter {
+  #findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+
+  #getNetworkClientById: NetworkController['getNetworkClientById'];
+
+  #subscriptions: SubscriptionEntry[] = [];
+
+  constructor(options: MultichainSubscriptionManagerOptions) {
+    super();
+    this.#findNetworkClientIdByChainId = options.findNetworkClientIdByChainId;
+    this.#getNetworkClientById = options.getNetworkClientById;
+  }
+
+  onNotification(
+    { scope, origin, tabId }: SubscriptionKey,
+    { method, params }: SubscriptionNotificationEvent,
+  ) {
+    this.emit('notification', origin, tabId, {
+      method: 'wallet_notify',
+      params: {
+        scope,
+        notification: { method, params },
+      },
+    });
+  }
+
+  #getSubscriptionEntry({
+    scope,
+    origin,
+    tabId,
+  }: SubscriptionKey): SubscriptionEntry | undefined {
+    return this.#subscriptions.find((subscriptionEntry) => {
+      return (
+        subscriptionEntry.scope === scope &&
+        subscriptionEntry.origin === origin &&
+        subscriptionEntry.tabId === tabId
+      );
+    });
+  }
+
+  #removeSubscriptionEntry({ scope, origin, tabId }: SubscriptionEntry) {
+    this.#subscriptions = this.#subscriptions.filter((subscriptionEntry) => {
+      return (
+        subscriptionEntry.scope !== scope ||
+        subscriptionEntry.origin !== origin ||
+        subscriptionEntry.tabId !== tabId
+      );
+    });
+  }
+
+  subscribe(subscriptionKey: SubscriptionKey) {
+    const subscriptionEntry = this.#getSubscriptionEntry(subscriptionKey);
+    if (subscriptionEntry) {
+      return subscriptionEntry.subscriptionManager;
+    }
+
+    const networkClientId = this.#findNetworkClientIdByChainId(
+      toHex(parseCaipChainId(subscriptionKey.scope as CaipChainId).reference),
+    );
+    const networkClient = this.#getNetworkClientById(networkClientId);
+    const subscriptionManager = createSubscriptionManager({
+      blockTracker: networkClient.blockTracker,
+      provider: networkClient.provider,
+    });
+
+    subscriptionManager.events.on(
+      'notification',
+      (message: SubscriptionNotificationEvent) => {
+        this.onNotification(subscriptionKey, message);
+      },
+    );
+
+    const newSubscriptionManagerEntry = {
+      ...subscriptionKey,
+      subscriptionManager,
+    };
+    subscriptionManager.destroy = subscriptionManager.middleware.destroy;
```

Comment:
It seems odd to me that one class is meddling with something that another class did. I am not sure I fully understand why we are doing this.

Reply:
Basically my goal here was to make it so that if `subscriptionManager.middleware.destroy()` is called that both the original logic of `subscriptionManager.middleware.destroy()` is triggered AND that `this.#unsubscribe()` cleanup is also called (that way the caller of `this.subscribe()` doesn't also have to manually clean this subscription up by making a `unsubscribeByXYZ()` call when the json-rpc-engine pipeline dies)

Unfortunately, the way that [SubscriptionManager sets `destroy` on the middleware is odd](https://github.com/MetaMask/eth-json-rpc-filters/blob/main/subscriptionManager.js#L24). This forces me to store that anonymous function elsewhere so that I can still call it after wrapping/replacing `subscriptionManager.middleware.destroy` to ensure it also calls `this.#unsubscribe`